packages/client-sdk-nodejs/src/internal/data-client.ts

```diff
@@ -158,6 +159,69 @@ export class DataClient implements IDataClient {
       this.configuration.getMiddlewares()
     );
   }
+  public connect(): Promise<void> {
+    const deadline = new Date();
+    deadline.setSeconds(deadline.getSeconds() + 10);
+
+    return this.connectWithinDeadline(deadline);
+  }
+
+  private connectWithinDeadline(deadline: Date): Promise<void> {
+    return new Promise((resolve, reject) => {
+      // Get the current state and initiate a connection
+      const currentState = this.clientWrapper
+        .getClient()
+        .getChannel()
+        .getConnectivityState(true);
+
+      this.logger.debug(`Client connectivity state: ${currentState}`);
+
+      if (currentState === ConnectivityState.READY) {
+        resolve();
+        return;
+      }
+
+      const now = new Date();
+
+      if (now >= deadline) {
+        this.logger.error('Unable to connect to Momento: deadline exceeded.');
+        resolve();
+        return;
+      }
+
+      this.clientWrapper
+        .getClient()
+        .getChannel()
+        .watchConnectivityState(currentState, deadline, (error?: Error) => {
+          if (error) {
+            this.logger.error(
+              `Unable to connect to Momento: ${error.name}. Please contact Momento if this persists.`
+            );
+            resolve();
+            return;
+          }
+
+          const newState = this.clientWrapper
+            .getClient()
+            .getChannel()
+            .getConnectivityState(false);
+
+          if (newState === ConnectivityState.READY) {
+            this.logger.debug(`Connected! Current state: ${newState}`);
+            resolve();
+          } else if (newState === ConnectivityState.CONNECTING) {
+            // The connection goes through the CONNECTING state before becoming READY,
+            // so we must watch it twice.
+            this.connectWithinDeadline(deadline).then(resolve).catch(reject);
```

Comment:
Does this warrant a delay before calling again? I believe there's a certain range (some 10s of millis) before the connection will transition to READY, and I'm not sure how often we may end up calling this recursively within that range. My concern is mostly with the call stack growing "too big", which ofcourse I don't know how big it can grow to be an issue in the first place.

Reply:
I've only seen the recursive call happen once. I think, since the watch only returns when a state change happens, it can only happen once.